[SPARK-10863][SPARKR] Method coltypes() to get R's data types of a DataFrame

olarayej opened the pull request:
Method coltypes() to get R's data types of a DataFrame

Commit:
SPARK-10807. Added as.data.frame as a synonym for collect().

Commit:
Removed operator %++%, which is a synonym for paste()

Commit:
Removed extra blank space.

Commit:
Removed extra spaces to comply with R style

Commit:
Moved setGeneric declaration to generics.R.
Changed setMethod stle
Removed return()

Commit:
Added test cases for as.data.frame

Commit:
Merge remote-tracking branch 'origin/SPARK-10807' into SPARK-10807

Conflicts:
	R/pkg/R/DataFrame.R

Commit:
Changed setMethod declaration to comply with standard

Commit:
Removed changes to .gitignore

Commit:
Merge remote-tracking branch 'upstream/master'

# Conflicts:
#	R/pkg/R/DataFrame.R

Commit:
coltypes

Commit:
coltypes

Commit:
coltypes

AmplabJenkins commented:
Can one of the admins verify this patch?

shivaram commented:
Jenkins, add to whitelist

shivaram commented:
Jenkins, ok to test

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43260 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43260/consoleFull) for   PR 8984 at commit [`b44152e`](https://github.com/apache/spark/commit/b44152ea6294cb3caf0cf2f92660b690693beaf7).

SparkQA commented:
[Test build #43260 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43260/console) for   PR 8984 at commit [`b44152e`](https://github.com/apache/spark/commit/b44152ea6294cb3caf0cf2f92660b690693beaf7).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43260/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1881,3 +1881,31 @@ setMethod("as.data.frame",
+            types <- as.character(t(as.data.frame(dtypes(x))[2, ]))
+
+            # Map Spark data types into R's data types
+            as.character(DATA_TYPES[types])
```
could you check for the case when it doesn't match the known types?

olarayej replied:
@felixcheung Yeah, that's a good point. I'm thinking coltypes() should always have an equivalent R data type for each column. We don't want method coltypes() to return NA's or throw an unsupported-type error cuz that would mean that the input DataFrame is inconsistent.

Therefore, it'd be just a matter of putting in DATA_TYPES, the list all possible values returned by dtypes() (If I'm missing any). I couldn't find that in the docs. Could you point me to the list?

Finally, I think the check for unsupported data types should be done instead in the coltypes()<- method and in the DataFrame initialization. coltypes() assumes the input DataFrame was assigned valid data types, which makes sense to me.

olarayej replied:
@felixcheung, @shivaram: Any thoughts on this one?

shivaram replied:
http://spark.apache.org/docs/latest/sql-programming-guide.html#data-types is a list that might be helpful.

Also I think it might make sense to try and map them to R types and if we fail to find a relevant one we fallback to the SparkSQL type

olarayej replied:
@shivaram I agree. I could use the mapping below (got the short types from schema.R:118):
scala  ->  R
"string"="character",
"long"="integer",
"short"="integer",
"integer"="integer"
"byte"="integer",
"double"="numeric",
"float"="numeric",
"decimal"="numeric",     
"boolean"="logical"

In any other case, I will use the same scala type. Sounds good?

shivaram replied:
Yep. This sounds good.

shivaram commented:
@olarayej Could you bring this PR up to date with master branch ?

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #43418 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43418/consoleFull) for   PR 8984 at commit [`523bfbf`](https://github.com/apache/spark/commit/523bfbf3b81a4bc8d7b45f2b382c11aab41976dc).

olarayej commented:
@shivaram Could you share the best practices to merge the changes from the master branch into the PR branch? This looks like a very common thing and the team (@NarineK, @adrian555, and myself) have tried quite a few options already, but none of them look pretty. We'd really appreciate any guidance. Thanks!

shivaram commented:
There are a number of ways to do this, so this is just the way I do it personally. In my case I have two remotes in my git setup. So my .git/config looks something like 
```
...
[remote "origin"]
        url = https://github.com/shivaram/spark-1.git
        fetch = +refs/heads/*:refs/remotes/origin/*
[remote "apache-spark"]
        url = https://github.com/apache/spark.git
        fetch = +refs/heads/*:refs/remotes/apache-spark/*
...
```

So if I'm on a feature branch say `SPARK-10863` I do the following
```
> git fetch apache-spark master 
...
From https://github.com/apache/spark
 * branch            master     -> FETCH_HEAD
...
> git merge FETCH_HEAD
... Accept the merge commit message that shows up
> git log -2 # Optionally use this to verify if things look fine
> git push origin SPARK-10863
... This will push changes to your fork for this branch
```

Let me know if this works for you

SparkQA commented:
[Test build #43418 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43418/console) for   PR 8984 at commit [`523bfbf`](https://github.com/apache/spark/commit/523bfbf3b81a4bc8d7b45f2b382c11aab41976dc).
 * This patch **fails PySpark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43418/
Test FAILed.

olarayej commented:
@shivaram Yes, that was helpful. Thank you! I have done the merge already. Jenkins, could you run tests?

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43456 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43456/consoleFull) for   PR 8984 at commit [`b1afe8e`](https://github.com/apache/spark/commit/b1afe8e39a9d568fea42d21ba1200bdc5d8b56b6).

SparkQA commented:
[Test build #43456 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43456/console) for   PR 8984 at commit [`b1afe8e`](https://github.com/apache/spark/commit/b1afe8e39a9d568fea42d21ba1200bdc5d8b56b6).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43456/
Test PASSed.

felixcheung commented:
+1 on @shivaram comment on data-type above.

Commit:
Added more types. Scala types that can't be mapped to R will remain as they are.

AmplabJenkins commented:
Merged build triggered.

Commit:
Removed white space

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43487 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43487/consoleFull) for   PR 8984 at commit [`76fe59a`](https://github.com/apache/spark/commit/76fe59a3ed0705111b1ad659e4e562d94786e96e).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43485/
Test FAILed.

SparkQA commented:
[Test build #43487 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43487/console) for   PR 8984 at commit [`76fe59a`](https://github.com/apache/spark/commit/76fe59a3ed0705111b1ad659e4e562d94786e96e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43487/
Test PASSed.

olarayej commented:
Thanks, @felixcheung @shivaram. I have committed my changes and tests have passed :-)

shivaram commented on R/pkg/inst/tests/test_sparkSQL.R:
```diff
@@ -1355,13 +1355,18 @@ test_that("SQL error message is returned from JVM", {
 
+test_that("Method coltypes() to get R's data types of a DataFrame", {
+  expect_equal(coltypes(irisDF), c(rep("numeric", 4), "character"))
+})
```
Could you add a test with some other types ? Also another one which runs into the `NA` case and uses the SQL type would be useful.

Commit:
Added more tests

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Fixed typo

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43495 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43495/consoleFull) for   PR 8984 at commit [`d53e8b3`](https://github.com/apache/spark/commit/d53e8b384838bb6bfc6ac2ab2bc80589ebd691fe).

SparkQA commented:
[Test build #43495 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43495/console) for   PR 8984 at commit [`d53e8b3`](https://github.com/apache/spark/commit/d53e8b384838bb6bfc6ac2ab2bc80589ebd691fe).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43495/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43494/
Test FAILed.

Commit:
Fixed typo

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43498 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43498/consoleFull) for   PR 8984 at commit [`baec23f`](https://github.com/apache/spark/commit/baec23f16419ed0d87d5964987ac504a7bad56e1).

SparkQA commented:
[Test build #43498 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43498/console) for   PR 8984 at commit [`baec23f`](https://github.com/apache/spark/commit/baec23f16419ed0d87d5964987ac504a7bad56e1).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43498/
Test FAILed.

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #43499 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43499/consoleFull) for   PR 8984 at commit [`baec23f`](https://github.com/apache/spark/commit/baec23f16419ed0d87d5964987ac504a7bad56e1).

SparkQA commented:
[Test build #43499 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43499/console) for   PR 8984 at commit [`baec23f`](https://github.com/apache/spark/commit/baec23f16419ed0d87d5964987ac504a7bad56e1).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43499/
Test FAILed.

shivaram commented:
@shaneknapp This one seems to be failing with 
```
sbt.ResolveException: download failed: org.apache.spark#spark-unsafe_2.10;1.5.0!spark-unsafe_2.10.jar
```
Any idea whats up ?

sun-rui commented on R/pkg/R/generics.R:
```diff
@@ -995,3 +995,7 @@ setGeneric("rbind", signature = "...")
+
+#' @rdname coltypes
+#' @export
+setGeneric("coltypes", function(x) standardGeneric("coltypes"))
```
style: { standardGeneric("coltypes") }

sun-rui commented:
I am thinking if we can put this function in a new separate R file other than DataFrame.R, as this function is R specific, and does not have a corresponding Scala version. This new file can also contain the function implemented in https://issues.apache.org/jira/browse/SPARK-10979

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1880,4 +1880,46 @@ setMethod("as.data.frame",
+                            "float"="numeric",
+                            "decimal"="numeric",
+                            "boolean"="logical"
+            )
```
You only handle primitive types here, but no complex types, like Array, Struct and Map.

It would be better you can refactor the type mapping related code here and that in SerDe.

olarayej replied:
@sun-rui For complex types (Array/Struct/Map), I can't think of any mapping to R types. Therefore, as agreed with @felixcheung and @shivaram, these will remain the same. For example:

Original column types: ["string", "boolean", "map..."]
Result of coltypes(): ["character", "logical", "map..."]

shivaram replied:
@olarayej I think the fall back mechanism here is good. But @sun-rui makes another good point that it will be good to have one unified place where we do a mapping from R types to java types. Right now part of that is in serialize.R / deserialize.R

Could you see if there is some refactoring we could do for this to not be duplicated ?

olarayej replied:
@sun-rui @shivaram 
The notion of coltypes is actually spread in three files: schema.R, serialize.R, deserialize.R.

In file serialize.R, method writeType (see below) turns the full data type into a one-character string. Then, method readTypedObject (see below), uses this one-character type to read accordingly. I suspect this is because complex types could be like map<String,String>? 

In my opinion, it would be better to use the full data type, as opposed to the first letter (which could be especially confusing since we support data types starting with the same letter Date/Double, String/Struct). Also, having the full data type would allow for centralizing the data types in one place, though this would require some major changes

We could have mapping arrays:

PRIMITIVE_TYPES <- c("string"="character",
+                            "long"="integer",
+                            "tinyint"="integer",
+                            "short"="integer",
+                            "integer"="integer",
+                            "byte"="integer",
+                            "double"="numeric",
+                            "float"="numeric",
+                            "decimal"="numeric",
+                            "boolean"="logical"

COMPLEX_TYPES  <- c("map", "array", "struct", ...)

DATA_TYPES <- c(PRIMITIVE_TYPES, COMPLEX_TYPES)


And then we'd need to modify deserialize.R, serialize.R, and schema.R to acknowledge these accordingly.

Thoughts?

writeType <- function(con, class) {
  type <- switch(class,
                 NULL = "n",
                 integer = "i",
                 character = "c",
                 logical = "b",
                 double = "d",
                 numeric = "d",
                 raw = "r",
                 array = "a",
                 list = "l",
                 struct = "s",
                 jobj = "j",
                 environment = "e",
                 Date = "D",
                 POSIXlt = "t",
                 POSIXct = "t",
                 stop(paste("Unsupported type for serialization", class)))
  writeBin(charToRaw(type), con)
}

readTypedObject <- function(con, type) {
  switch (type,
    "i" = readInt(con),
    "c" = readString(con),
    "b" = readBoolean(con),
    "d" = readDouble(con),
    "r" = readRaw(con),
    "D" = readDate(con),
    "t" = readTime(con),
    "a" = readArray(con),
    "l" = readList(con),
    "e" = readEnv(con),
    "s" = readStruct(con),
    "n" = NULL,
    "j" = getJobj(readString(con)),
    stop(paste("Unsupported type for deserialization", type)))
}

shivaram replied:
The single character names are to reduce the amount of data serialized when we transfer these data types to the JVM. Its not meant to be remembered by anybody so I don't see it being a source of confusion. @sun-rui also added tests which ensure these mappings don't break.

However I think having a list of primitive types, complex types and mapping in a common file (types.R ?) sounds good to me.

olarayej commented:
@sun-rui @shivaram Sure, I can move the code to a new file if needed. What should I call this file?

felixcheung commented:
datatype.R? :)

Commit:
Moved coltypes to new file types.R and refactored schema.R

Commit:
Updated DESCRIPTION file to add types.R

Commit:
Updated DESCRIPTION file

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
[Test build #43939 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43939/consoleFull) for   PR 8984 at commit [`49f2393`](https://github.com/apache/spark/commit/49f239305a8c0c944ab15cafe4eb71ced30f5c4b).

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

olarayej commented:
@shivaram Fair enough :-). Then, it looks to me that there's not much of a refactoring I can do in deserialize.R/serialize.R. However, I did do a small change in schema.R to centralize the primitive data types and their mapping to R's. I also moved my code to types.R as discussed. Thanks!

SparkQA commented:
**[Test build #43941 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43941/consoleFull)** for PR 8984 at commit [`d3d9933`](https://github.com/apache/spark/commit/d3d9933ab7e7cef9de02b09f4ebb0bf6eedbf7df).

SparkQA commented:
**[Test build #43941 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43941/consoleFull)** for PR 8984 at commit [`d3d9933`](https://github.com/apache/spark/commit/d3d9933ab7e7cef9de02b09f4ebb0bf6eedbf7df).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43941/
Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43945 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43945/consoleFull)** for PR 8984 at commit [`1732a06`](https://github.com/apache/spark/commit/1732a061bd29ef4e95d49b092630a3b0cd5f7fe1).

SparkQA commented:
**[Test build #43945 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43945/consoleFull)** for PR 8984 at commit [`1732a06`](https://github.com/apache/spark/commit/1732a061bd29ef4e95d49b092630a3b0cd5f7fe1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43945/
Test PASSed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1883,7 +1883,8 @@ setMethod("as.data.frame",
             collect(x)
-          })
+          }
+)
```
probably more consistent in style to have `})`

felixcheung commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,67 @@
+#' @param x (DataFrame)
+#' @return value (character) A character vector with the column types of the given DataFrame
+#' @rdname coltypes
+setMethod("coltypes",
```
I actually think while the types are here the `coltypes` function should be in DataFrame.R? Thought?

sun-rui replied:
yeah, As discussed in https://github.com/apache/spark/pull/9012, for now, let's keep coltypes() in DataFrame.R.

felixcheung commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,67 @@
+          function(x) {
+            # Get the data types of the DataFrame by invoking dtypes() function.
+            # Some post-processing is needed.
+            types <- as.character(t(as.data.frame(dtypes(x))[2, ]))
```
probably better for style consistency to have `[2,]` instead of `[2, ]`

SparkQA commented:
[Test build #43939 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43939/console) for   PR 8984 at commit [`49f2393`](https://github.com/apache/spark/commit/49f239305a8c0c944ab15cafe4eb71ced30f5c4b).
 * This patch **passes all tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43939/
Test PASSed.

sun-rui commented on R/pkg/R/generics.R:
```diff
@@ -1011,3 +1010,7 @@ setGeneric("as.data.frame")
+
+#' @rdname coltypes
+#' @export
+setGeneric("coltypes", function(x) standardGeneric("coltypes"))
```
coding style: { standardGeneric("coltypes") }

Commit:
Coding style for setGeneric definition

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Coding style

SparkQA commented:
**[Test build #43996 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43996/consoleFull)** for PR 8984 at commit [`22c27ea`](https://github.com/apache/spark/commit/22c27eafa421ee51b5876bf4992f6d568ab065f1).

Commit:
Coding style

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #43997 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43997/consoleFull)** for PR 8984 at commit [`76747bc`](https://github.com/apache/spark/commit/76747bca2103022ded4a5000bf8fa3e8ed055169).

SparkQA commented:
**[Test build #43997 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43997/consoleFull)** for PR 8984 at commit [`76747bc`](https://github.com/apache/spark/commit/76747bca2103022ded4a5000bf8fa3e8ed055169).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43997/
Test FAILed.

SparkQA commented:
**[Test build #43996 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43996/consoleFull)** for PR 8984 at commit [`22c27ea`](https://github.com/apache/spark/commit/22c27eafa421ee51b5876bf4992f6d568ab065f1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/43996/
Test PASSed.

olarayej commented:
@felixcheung @sun-rui Thanks for your comments! I have addressed them all. Anything else I should do?

felixcheung commented:
looks good

olarayej commented:
Nice! Could we merge this then?

sun-rui commented:
@olarayej, give me some time on reviewing the type mapping.

olarayej commented:
@sun-rui Sure! Thank you!

shivaram commented:
@sun-rui Friendly ping on this

felixcheung commented:
@shivaram as I was discussing with @sun-rui in #9218 - I think coltypes() could probably handle complex type (JVM -> R) by mapping "map<string,int>" -> "environment", "array<string>" -> "list" but this conversion is not perfect, at least from R -> JVM.

I can have this
```
> e <- new.env()
> e[["abd"]] <- 1276
> e[["84798"]] <- "abc"
> l <- list(e)
> df <- createDataFrame(sqlContext, list(l))
> df
DataFrame[_1:map<string,string>]
```

So `env` supports mixed type values they are mapped to `map<string, string>` on JVM.

In fact, this DataFrame doesn't seem to work properly
```
> head(df)
15/10/27 19:02:02 ERROR Executor: Exception in task 0.0 in stage 6.0 (TID 6)
scala.MatchError: 1276.0 (of class java.lang.Double)
	at org.apache.spark.sql.catalyst.CatalystTypeConverters$StringConverter$.toCatalystImpl(CatalystTypeConverters.scala:295)
```

shivaram commented:
Is this the same as https://issues.apache.org/jira/browse/SPARK-11281 ?

felixcheung commented:
The error looks different, possibly related but not exactly same cause

sun-rui commented:
For a DataFrame column of MapType, the key type and value type must be consistent.

> e[["abd"]] <- 1276
> e[["84798"]] <- "abc"
In your example, the first value type  is numeric, while the second value type is character.

I think this issue is same as https://issues.apache.org/jira/browse/SPARK-11167

sun-rui commented:
A rough idea is that:
    for coltypes(), "ArrayType" -> "list", "MapType" -> "environment", "StructType" -> "struct" (struct is a named list with "struct" class attribute)
    for coltypes<-(), throws "unsupported" error.

felixcheung commented:
I'm concerned with the lack of "reversibility" as in `coltypes(x) <- coltypes(x)`
Could I propose expanding on your suggestion to return `na` for non-atomic type?
`coltypes<-()` can skip over `na`
`coltypes()` can have an optional argument default atomic.only = TRUE - which can be set to FALSE to get `list`, `environment`, `struct` etc  for non-atomic vector types.

sun-rui commented:
Per the Scala API doc on cast() of Column, the supported types are: string, boolean, byte, short, int, long, float, double, decimal, date, timestamp. That is, complext types are not supported as a target type. So for coltypes<-(), regardless of the input (NA or not), it should not cast the type of a column of a complex type. If the corresponding input is NA, coltype<-() can silently skip the column, while if not NA, then prompts a warning.

What I am concerned about is that coltypes<-() actually returns a new DataFrame instead in-place changing of the schema of the DataFrame (which is not supported by Spark Core). Is this a desired behavior?

felixcheung commented:
@sun-rui the implementation of `coltypes<-` is in-place changing the DataFrame? You can see the test code at https://github.com/apache/spark/pull/9218/files#diff-56a2e82b105ec1dd2374b9653fa5fa2aR615

olarayej commented:
@sun-rui @felixcheung In types.R, should I then replace:

COMPLEX_TYPES <- c("map"=NA, "array"=NA, "struct"=NA)

by:

COMPLEX_TYPES <- c("map"="environment", "array"="list", "struct"="struct")

felixcheung commented:
@olarayej from what I gather from our conversation, "map" = NA should be what we want initially. @sun-rui ?

felixcheung commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+
+# The primitive data types, where names(PRIMITIVE_TYPES) are Scala types whereas
+# values are equivalent R types.
+PRIMITIVE_TYPES <- c(
```
nit: you might want to change this to environment for fixed-cost / O(1) lookup

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1914,3 +1914,29 @@ setMethod("attach",
+          function(x) {
+            # Get the data types of the DataFrame by invoking dtypes() function.
+            # Some post-processing is needed.
+            types <- as.character(t(as.data.frame(dtypes(x))[2,]))
```
the code can be simplified as:
```
as.character(lapply(x, function(t) { t[[2]] }))
```

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+# The primitive data types, where names(PRIMITIVE_TYPES) are Scala types whereas
+# values are equivalent R types.
+PRIMITIVE_TYPES <- c(
+  "byte"="integer",
```
"byte" should be "tinyint"

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+# values are equivalent R types.
+PRIMITIVE_TYPES <- c(
+  "byte"="integer",
+  "tinyint"="integer",
```
"smallint"="integer",

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+PRIMITIVE_TYPES <- c(
+  "byte"="integer",
+  "tinyint"="integer",
+  "integer"="integer",
```
"int"="integer",

sun-rui replied:
"bigint"="numeric"

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "integer"="integer",
+  "float"="numeric",
+  "double"="numeric",
+  "numeric"="numeric",
```
"decimal"="numeric",

felixcheung replied:
I'm concern with this. decimal does not map exactly to numeric
And it says it is not supported:
http://spark.apache.org/docs/latest/sql-programming-guide.html#data-types

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "float"="numeric",
+  "double"="numeric",
+  "numeric"="numeric",
+  "character"="character",
```
remove this line

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "numeric"="numeric",
+  "character"="character",
+  "string"="character",
+  "binary"=NA,
```
"binary"="raw"

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "character"="character",
+  "string"="character",
+  "binary"=NA,
+  "raw"=NA,
```
remove this line

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "string"="character",
+  "binary"=NA,
+  "raw"=NA,
+  "logical"="logical",
```
remove this line

sun-rui replied:
remove this line

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "raw"=NA,
+  "logical"="logical",
+  "boolean"="logical",
+  "decimal"="numeric",
```
remove this line as duplicated as above

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "logical"="logical",
+  "boolean"="logical",
+  "decimal"="numeric",
+  "timestamp"=NA,
```
"timestamp"="POSIXct",

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,41 @@
+  "boolean"="logical",
+  "decimal"="numeric",
+  "timestamp"=NA,
+  "date"=NA)
```
"date"="Date"

Commit:
Fixed data type mapping. Put data types in an environment for more efficient lookup

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

Commit:
Removed unnecessary cat

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44947 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44947/consoleFull)** for PR 8984 at commit [`f418f84`](https://github.com/apache/spark/commit/f418f848e91886a4dcd4484a32a7cc3e6459a593).

SparkQA commented:
**[Test build #44947 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44947/consoleFull)** for PR 8984 at commit [`f418f84`](https://github.com/apache/spark/commit/f418f848e91886a4dcd4484a32a7cc3e6459a593).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44947/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44946/
Test FAILed.

Commit:
Removed white space

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44949/
Test FAILed.

olarayej commented:
I'm seeing the error below. Jenkins, could you rerun the tests? Thanks!

ERROR: Error fetching remote repo 'origin'

shivaram commented:
Jenkins, retest this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44953 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44953/consoleFull)** for PR 8984 at commit [`f333173`](https://github.com/apache/spark/commit/f333173b52f31050e29d792e92ace2aba91be7be).

SparkQA commented:
**[Test build #44953 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44953/consoleFull)** for PR 8984 at commit [`f333173`](https://github.com/apache/spark/commit/f333173b52f31050e29d792e92ace2aba91be7be).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44953/
Test FAILed.

Commit:
Removed blank space

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44955 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44955/consoleFull)** for PR 8984 at commit [`63090f6`](https://github.com/apache/spark/commit/63090f68620510578957025249c1b6fac72d9c42).

SparkQA commented:
**[Test build #44955 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44955/consoleFull)** for PR 8984 at commit [`63090f6`](https://github.com/apache/spark/commit/63090f68620510578957025249c1b6fac72d9c42).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44955/
Test FAILed.

Commit:
Update DataFrame.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44962 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44962/consoleFull)** for PR 8984 at commit [`cf068a0`](https://github.com/apache/spark/commit/cf068a01da4d7cce08290159a99890443caa124e).

SparkQA commented:
**[Test build #44962 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44962/consoleFull)** for PR 8984 at commit [`cf068a0`](https://github.com/apache/spark/commit/cf068a01da4d7cce08290159a99890443caa124e).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44962/
Test FAILed.

felixcheung commented on R/pkg/R/DataFrame.R:
```diff
@@ -1914,3 +1914,34 @@ setMethod("attach",
+
+            # Map Spark data types into R's data types using DATA_TYPES environment
+            rTypes <- lapply(types, function(x) {
+              if (exists(x, envir=DATA_TYPES)) {
```
nit: you could probably do
```
type <- DATA_TYPES[[x]]
if (is.null(x)) {
   stop("error")
}
```

Commit:
Update types.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44977 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44977/consoleFull)** for PR 8984 at commit [`24976fb`](https://github.com/apache/spark/commit/24976fb7ae222507f28beec47e77ef8b1136fa9a).

SparkQA commented:
**[Test build #44977 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44977/consoleFull)** for PR 8984 at commit [`24976fb`](https://github.com/apache/spark/commit/24976fb7ae222507f28beec47e77ef8b1136fa9a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44977/
Test FAILed.

Commit:
Update types.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #44986 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44986/consoleFull)** for PR 8984 at commit [`df6606b`](https://github.com/apache/spark/commit/df6606b8832c9979df87c4c878e098a59cb04707).

SparkQA commented:
**[Test build #44986 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44986/consoleFull)** for PR 8984 at commit [`df6606b`](https://github.com/apache/spark/commit/df6606b8832c9979df87c4c878e098a59cb04707).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/44986/
Test FAILed.

sun-rui commented on R/pkg/R/types.R:
```diff
@@ -0,0 +1,43 @@
+  "date"="Date"))
+
+# The complex data types. These do not have any direct mapping to R's types.
+COMPLEX_TYPES <- as.environment(list(
```
complex types form dtypes() are not simply "map", "array", or "struct", but including type information of its elements. So you have to check types by something like "startsWith("map")"...

felixcheung replied:
good call, i guess we should add a test with complex type

Commit:
Update DataFrame.R

Commit:
Added tests for complex types

Commit:
Update types.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45045 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45045/consoleFull)** for PR 8984 at commit [`0d30dc1`](https://github.com/apache/spark/commit/0d30dc1057f87895444705b7cd7fb5bc808ce15b).

SparkQA commented:
**[Test build #45045 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45045/consoleFull)** for PR 8984 at commit [`0d30dc1`](https://github.com/apache/spark/commit/0d30dc1057f87895444705b7cd7fb5bc808ce15b).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45045/
Test FAILed.

felixcheung commented:
I'm a bit confused - I thought `map` `array` `struct` should return NA?

olarayej commented:
As per @sun-rui comment, I thought array would be mapped to list. I can change it back to NA too.

Commit:
Update types.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45050 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45050/consoleFull)** for PR 8984 at commit [`7af7d4e`](https://github.com/apache/spark/commit/7af7d4eb80da243559317c235a87fb73abc435ed).

SparkQA commented:
**[Test build #45050 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45050/consoleFull)** for PR 8984 at commit [`7af7d4e`](https://github.com/apache/spark/commit/7af7d4eb80da243559317c235a87fb73abc435ed).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45050/
Test FAILed.

Commit:
Update test_sparkSQL.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45052/consoleFull)** for PR 8984 at commit [`f6fa695`](https://github.com/apache/spark/commit/f6fa695413a26d6aa70894af36bbf9bb8fe7f711).

SparkQA commented:
**[Test build #45052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45052/consoleFull)** for PR 8984 at commit [`f6fa695`](https://github.com/apache/spark/commit/f6fa695413a26d6aa70894af36bbf9bb8fe7f711).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45052/
Test PASSed.

olarayej commented:
Thanks @sun-rui , @felixcheung for all your comments. I think I've addressed them all, including:
- Added tests for complex types
- Used environments for more efficient look up of data types
- Fixed primitive/complex data types mapping

Hope this covers everything. Please let me know! Thank you! :)

felixcheung commented:
looks good. thanks

shivaram commented:
Thanks @olarayej and @felixcheung  -- @sun-rui any other comments ?

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1914,3 +1914,46 @@ setMethod("attach",
+              type <- PRIMITIVE_TYPES[[x]]
+              if (is.null(type)) {
+                # Check for complex types
+                for (t in names(COMPLEX_TYPES)) {
```
Try to eliminate for loop in R, could you try something like:
 Filter(function(t) {substring(x, 1, nchar(t)) == t}, names(COMPLEX_TYPES))

felixcheung replied:
Or
```
Filter(function(t) { grep("start", txt) == 1 }, names(COMPLEX_TYPES))
```
?

sun-rui commented:
I am still not clear about what appropriate  R types are returned for complex types, so for now, let's first return NA for them. We can address them later.

Commit:
Removed for loop

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45132 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45132/consoleFull)** for PR 8984 at commit [`67cb102`](https://github.com/apache/spark/commit/67cb102a9e9b36efe96dda89c2fa5d50b0e90d22).

SparkQA commented:
**[Test build #45132 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45132/consoleFull)** for PR 8984 at commit [`67cb102`](https://github.com/apache/spark/commit/67cb102a9e9b36efe96dda89c2fa5d50b0e90d22).
 * This patch **fails some tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45132/
Test FAILed.

Commit:
Update DataFrame.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45138 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45138/consoleFull)** for PR 8984 at commit [`f4b0146`](https://github.com/apache/spark/commit/f4b014612200117c6abd33f0c19552fe3a0188b9).

SparkQA commented:
**[Test build #45138 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45138/consoleFull)** for PR 8984 at commit [`f4b0146`](https://github.com/apache/spark/commit/f4b014612200117c6abd33f0c19552fe3a0188b9).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45138/
Test FAILed.

Commit:
Update DataFrame.R

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45151 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45151/consoleFull)** for PR 8984 at commit [`5c3e3fa`](https://github.com/apache/spark/commit/5c3e3fae3b6f6bee1c69ffbb0f02c2922bbcd9a1).

SparkQA commented:
**[Test build #45151 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45151/consoleFull)** for PR 8984 at commit [`5c3e3fa`](https://github.com/apache/spark/commit/5c3e3fae3b6f6bee1c69ffbb0f02c2922bbcd9a1).
 * This patch **fails R style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45151/
Test FAILed.

Commit:
Removed blank space

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
**[Test build #45154 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45154/consoleFull)** for PR 8984 at commit [`6794d82`](https://github.com/apache/spark/commit/6794d8262839787b217746030de4ef965cae72e4).

sun-rui commented on R/pkg/R/DataFrame.R:
```diff
@@ -1914,3 +1914,43 @@ setMethod("attach",
+#' 
+#' @name coltypes
+#' @title Get column types of a DataFrame
+#' @param x (DataFrame)
```
Could you update the style of  function description to be more consistent with other existing ones?

felixcheung replied:
I can change this when updating my PR #9218

sun-rui commented:
LGTM, except a comment on function description style.

SparkQA commented:
**[Test build #45154 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45154/consoleFull)** for PR 8984 at commit [`6794d82`](https://github.com/apache/spark/commit/6794d8262839787b217746030de4ef965cae72e4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45154/
Test PASSed.

felixcheung commented:
@olarayej you would need to rebase add the conflict before this can be merged.

Commit:
Merged

Commit:
Merge tests and description files

olarayej commented:
@felixcheung I have tried quite a few things already but unfortunately, I haven't been able to do the rebase. Could you provide some suggestions? Thanks!

shivaram commented:
@olarayej Do the git merge commands in https://github.com/apache/spark/pull/8984#issuecomment-146695587 not work ?

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
**[Test build #45398 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45398/consoleFull)** for PR 8984 at commit [`0bc5b35`](https://github.com/apache/spark/commit/0bc5b35220aabb96930190bfd6b17ff97ea2cb9b).

olarayej commented:
@shivaram @felixcheung 
I followed the same steps described by @shivaram.

What's confusing for us is that every time we run a fetch followed by a merge, it triggers conflicts with a number of files that we haven't modified (even outside the R folder). After I solved all conflicts, and ran a push, it also pushed those files. Now there are 194 modified files, which makes things pretty messy.

I'm thinking about creating a new branch and discard this one. Thoughts?

SparkQA commented:
**[Test build #45398 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45398/consoleFull)** for PR 8984 at commit [`0bc5b35`](https://github.com/apache/spark/commit/0bc5b35220aabb96930190bfd6b17ff97ea2cb9b).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Build triggered.

AmplabJenkins commented:
Build started.

SparkQA commented:
**[Test build #45400 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45400/consoleFull)** for PR 8984 at commit [`ba091fb`](https://github.com/apache/spark/commit/ba091fbc1a9e9aad12d48c78cef792dbab69ef52).

SparkQA commented:
**[Test build #45400 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/45400/consoleFull)** for PR 8984 at commit [`ba091fb`](https://github.com/apache/spark/commit/ba091fbc1a9e9aad12d48c78cef792dbab69ef52).
 * This patch **fails to build**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

shivaram commented:
Yeah something seems to be messed up. You shouldn't get other files changed if you do a fetch + merge as long as the rest of your tree is synced to the same place. 

You can open a new PR if you feel that its getting messy in this case -- The only downside is that we lose all these comments we had etc. but since this PR is close to being merged its probably fine in this case.

olarayej commented:
I have created a new branch and PR #9579 to follow up on this.

shivaram commented:
@olarayej Could you close this PR ? Only the person who opened the PR can close it and it helps clear our PR queue at https://spark-prs.appspot.com/#r

olarayej commented:
Closing this PR as #9579 has been created to follow up....
